README
ntalk - Ncurses Talk
====================

 REQUIREMENTS
 This version of ntalk-client requires ntalk-server 1.0.0 or later.
 
 DESCRIPTION
 
 Ntalk is a chat application. Is is designed to be much more
 friendly than Unix talk and provide nice console interface.
 You will need ntalks (ntalk server) to run ntalk. You
 can download it from ntalk homepage (see below).
 I also provide Kntalk, a KDE ntalk client for the X Window
 System. Visit my web page for details.
 
 CONFIGURATION
 
 Ntalk uses .ntalkrc configuration file. If you already used
 ntalk, you can find it in your home directory. Otherwise,
 a sample one is provided with this package. Some options
 are saved by ntalk itself (nick, server address), other
 you have to change by editing your configuration file.
 Option 'body' (value) is placed after the '=' sign. 
 You should not change the data before that sign, because
 ntalk won't be able to read it. But if you do so, it won't
 cause crash (or other problem), the option will be just
 skipped.

 INFORMATION

 Author   : Krzysztof Czuba <kab@interia.pl>
 Homepage : http://sourceforge.net/projects/ntalk/
 
 This program is free software which I release
 under the GNU General Public License. Read the
 COPYING file for more info.
  
 Bug reports, suggestions, patches are welcome.
 Please send them to kab@interia.pl.
 For information about instalation read INSTALL.


client/CHANGELOG
1.0.0
-----
* Minor fixes so that it will compile with modern compilers.

 0.6.3
-------

* very minor fixes
* updated documentation

 0.6.2
-------

* improved configuration file handling. It is not
  overwritten now (just modified), so you can add
  your comments and they won't be deleted
* added configurable color schemes
* added configurable leave/away messages
* added sample configuration file to package
* updated README file

 0.6.1
-------

* fixed crash after server shutdown
* added command "/sclear"

server:
* fixed bug which cause incorrect response
  fot 'online' command
* core improvements

 0.6
-----

* completly rewritten in C++
* now using C++ string class
* new communication model (similar to IRC)
* new command (/whois) added
* improved configuration file handling
* gethostbyname() support added
* INSTALL update
* UI fixes

server:
* completly rewritten in C++
* new communication model
* completly new user management system based on lists

 0.4.5
-------

* added window with online users' nicknames list
  (For now it is not configurable, so if you  don't 
  like it just use 0.4.2 version)
* removed option '/online', because it's
  "automatic" now :)

 0.4.2
-------

* removed usleep instruction, so the screen is updated 
  immediately (it was updated every 0.5 second before)
* #define VERSION ".." added 
* updated help, because I have implemented command /about, 
  but forgot to add it to command list :)
* corrected default server address (127.0.0.1)
* updated README file

 0.4.1
-------

* removed change_nick() function, so you can't change
  your nick during the session. I decided to remove
  this function, because it caused ugly bug, which
  was difficult to fix for me.
* little code cleaning
* updated documentation

 0.4
-----

* added "-lsocket -lnsl" to the Makefile, so
  that the program compiles under Solaris (th.fischer@gmx.net)
* added target "clean" to Makefile (th.fischer@gmx.net)  
* changed "ncurses/ncurses.h" to "ncurses.h"
* now ntalk uses uid for initial nick (kris@unix.sbg.ac.at)
* 'security' option - 2 users can't connect with
  the same nickname
* /online command added. Use it to check who's online
* 'away' mode added
* config file created; nick and server address
  are now saved in ~/.ntalkrc file
* server: fixed small bug which causes, that when
  a client was killed server didn't decrease
  clients number

 0.3.1 
-------
(First public release)

* big bug fixed (input text doesn't seem to
  disappear now when the main 'window' is scrolling)
* added feature that shows how many users are online 
* general code cleaning

 0.3
-----

* added built-in mini-help
* 'syntax highlighting' improved
* added 3 new user commands:
  - /server which shows server version etc.
  - /time shows current time
  - /nick changes your nickname
* added join and leave notification ('j.smith joined at...')
* bugfixes


 0.2
-----

* changed one-char command names to full
* fixed parse_command()
* replaced fork() with threads, so its working correctly 
  now (one process writes on the screen)
* fixed 'syntax highlighting'
* small improvments (i.e you can't now enter too long or short nick)
* bugfixes...


server/ntalks.cpp
/*
 *  ntalks.cpp
 *
 *  This file is part of ntalk chat server distribution
 *  Copyright (C) 2001 by Krzysztof Czuba (kab@interia.pl)
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 */

#include <iostream>
#include <cstring>  	// C++ string
#include <list>
using namespace std;

#include <sys/types.h>
#include <sys/socket.h>
#include <netinet/in.h>
#include <sys/time.h>
#include <sys/ioctl.h>
#include <unistd.h>
#ifdef SOLARIS
#include <sys/filio.h>
#endif
#include <signal.h>

#define VERSION "1.0.0"

#define MESSAGE 	"01"
#define JOIN 		"02"
#define LEAVE 		"03"
#define DISCONNECT 	"04"
#define SHUTDOWN 	"05"
#define ONLINE 		"06"
#define AWAY 		"07"
#define AWAYOFF 	"08"
#define SERVER 		"09"
#define WHOIS 		"10"
#define SRVMESSAGE 	"11"


struct user {
    string nick;
    string addr;
    string since;
    int fd;
};

list<user> user_list;

int server_sockfd, client_sockfd;
int server_len, client_len;
int fd_count;
int result, x, online_count, already_sigint;
char msg_r[512], msg_s[512];

fd_set readfds, testfds;

sockaddr_in server_address, client_address;


void add_fd(int fd)
{
    user u;
    u.nick = "";
    u.fd = fd;

    user_list.push_back(u);
    fd_count++;
}

void del_fd(int fd)
{
    list<user>::iterator i;
    for(i = user_list.begin(); i != user_list.end(); ++i)
    {
	user u = *i;

	if(u.fd == fd)
	{
	    user_list.erase(i);
	    fd_count--;
	    return;
	}
    }
}

void add_nick(int fd, string nick, string addr, string since)
{
    list<user>::iterator i;
    for(i = user_list.begin(); i != user_list.end(); ++i)
    {
	if(i->fd == fd)
	{
	    i->nick = nick;
	    i->addr = addr;
	    i->since = since;
	    return;
	}
    }
}

bool check_nick(string nick)
{
    list<user>::const_iterator i;
    for(i = user_list.begin(); i != user_list.end(); ++i)
    {
	if(i->nick == nick) return 0;
    }

    return 1;
}

void send_all(string messg)
{
    strcpy(msg_s, messg.c_str());

    list<user>::const_iterator i;
    for(i = user_list.begin(); i != user_list.end(); ++i)
    {
	write(i->fd, &msg_s, sizeof(msg_s));
    }
}

void send_fd(const string messg, int fd)
{
    strcpy(msg_s, messg.c_str());
    write(fd, &msg_s, sizeof(msg_s));
}

user find_user(string nick)
{
    list<user>::const_iterator i;
    for(i = user_list.begin(); i != user_list.end(); ++i)
    {
	user u = *i;
	if(u.nick == nick) return u;
    }

    user us;
    us.fd = -1;
    return us;
}

string time()
{
    tm *tm;
    time_t now;
    char tt[9];
    string strtt;

    time(&now);
    tm = localtime(&now);
    strftime(tt, 9, "%H:%M:%S", tm);
    strtt = tt;

    return strtt;
}

string str_online()
{
    string str = "server@server:06:fd:";

    list<user>::iterator i;
    for(i = user_list.begin(); i != user_list.end(); ++i)
    {
	str.append(i->nick);
	str.append(";");
    }

    str.append("\n");

    return str;
}

string create_message_string(string com,
			     string data,
			     string sender = "server@server",
			     string to = "all")
{
    string m = sender;
    m.append(":" + com + ":");
    m.append(to + ":");
    m.append(data);
    m.append("\n");

    return m;
}

void parse_command(const string messg, int fd)
{
    string msg = messg;
    string nick, addr, com, to, data;

    // sender@host:01:all:message

    nick = msg.substr(0, msg.find('@'));
    msg.erase(0, msg.find('@') + 1);
    addr = msg.substr(0, msg.find(':'));
    msg.erase(0, msg.find(':') + 1);
    com = msg.substr(0, msg.find(':'));
    msg.erase(0, msg.find(':') + 1);
    to  = msg.substr(0, msg.find(':'));
    msg.erase(0, msg.find(':') + 1);
    data = msg.substr(0, msg.find('\n'));

    if(com.compare(MESSAGE) == 0)
    {
	send_all(messg);
    }
    else if(com.compare(JOIN) == 0)
    {
	if(!check_nick(nick))
	{
	    string str = create_message_string(DISCONNECT, "Nick already in use");
	    send_fd(str, fd);
	    return;
	}

	add_nick(fd, nick, addr, time());
	send_all(messg);
	send_all(str_online());
    }
    else if(com.compare(LEAVE) == 0)
    {
	send_all(messg);
    }
    else if(com.compare(SERVER) == 0)
    {
	string str = create_message_string(SERVER, "ntalk server " VERSION);
	send_fd(str, fd);
    }
    else if(com.compare(ONLINE) == 0)
    {
	string str = str_online();
	send_fd(str, fd);
    }
    else if(com.compare(AWAY) == 0)
    {
	send_all(messg);
    }
    else if(com.compare(AWAYOFF) == 0)
    {
	send_all(messg);
    }
    else if(com.compare(WHOIS) == 0)
    {
	user u = find_user(data);
	if(u.fd == -1)
	{
	    string tmpstr = data + ": No such user";
	    string str = create_message_string(WHOIS, tmpstr);
	    send_fd(str, fd);
	}
	else
	{
	    string tmpstr = u.nick + "@" + u.addr;
	    string str = create_message_string(WHOIS, u.since, tmpstr);
	    send_fd(str, fd);
	}
    }
}

static void signal_interrupt(int sig)
{
    string msg = create_message_string(SHUTDOWN, "Server shutdown");

    if (!already_sigint)
    {
	already_sigint = 1;
	send_all(msg);
	cout << "Server shut down" << endl;
	exit(0);
    }
}

int main(int argc, char *argv[])
{
    int opt;

    while((opt = getopt(argc, argv, "v")) != -1)
    {
	if(opt == 'v')
	{
	    cout << "version " VERSION "\n";
	    exit(0);
	}

	else exit(0);
    }

    server_sockfd = socket(AF_INET, SOCK_STREAM, 0);

    server_address.sin_family = AF_INET;
    server_address.sin_addr.s_addr = htonl(INADDR_ANY);
    server_address.sin_port = htons(9734);
    server_len = sizeof(server_address);

    bind(server_sockfd, (struct sockaddr *)&server_address, server_len);
    listen(server_sockfd, 5);

    online_count = 0;
    fd_count = 0;

    cout << "Running ntalk server " VERSION "\n";

    already_sigint = 0;
    signal(SIGINT, signal_interrupt);

    FD_ZERO(&readfds);
    FD_SET(server_sockfd, &readfds);
    while(1)
    {
	int fd;
	int nread;

	testfds = readfds;

	result = select(FD_SETSIZE, &testfds, (fd_set *)0,
			(fd_set *)0, (struct timeval *) 0);

	if(result < 1)
	{
	    cerr << "server: select";
	    exit(1);
	}

	for(fd = 0; fd < FD_SETSIZE; fd++)
	{
	    if(FD_ISSET(fd, &testfds))
	    {
		if(fd == server_sockfd)
		{
		    client_sockfd = accept(server_sockfd, (sockaddr *)&client_address, (socklen_t *)&client_len);
		    FD_SET(client_sockfd, &readfds);
		    online_count++;
		    add_fd(client_sockfd);
		}
		else
		{
		    ioctl(fd, FIONREAD, &nread);
		    if(nread == 0)
		    {
			close(fd);
			FD_CLR(fd, &readfds);
			online_count--;
			del_fd(fd);
			send_all(str_online());
		    }

		    else
		    {
			read(fd, &msg_r, sizeof(msg_r));
			string msgg = msg_r;
			parse_command(msgg, fd);
		    }
		}
	    }
	}
    }
}


client/ntalk.cpp
/*
 *  ntalk.cpp
 *
 *  This file is part of ntalk chat client distribution
 *  Copyright (C) 2001 by Krzysztof Czuba (kab@interia.pl)
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 */

#define _REENTRANT	/* required by pthread */

#include <iostream>
#include <fstream>
#include <cstring> 	// C++ string
#include <algorithm>
using namespace std;

#include <sys/types.h>
#include <sys/socket.h>
#include <pthread.h>
#include <netinet/in.h>
#include <arpa/inet.h>
#include <netdb.h>
#include <unistd.h>
#include <pwd.h>
#include <signal.h>
#include <ncurses.h>

#define VERSION "1.0.0"
#define MAX_DATA_SIZE 512

#define MESSAGE 	"01"
#define JOIN 		"02"
#define LEAVE 		"03"
#define DISCONNECT 	"04"
#define SHUTDOWN 	"05"
#define ONLINE 		"06"
#define AWAY 		"07"
#define AWAYOFF 	"08"
#define SERVER 		"09"
#define WHOIS 		"10"
#define SRVMESSAGE 	"11"


struct color {
    int you;
    int other;
    int nick;
    int bracket;
    int status;
};

struct message {
    string leave;
    string away;
};

WINDOW *input_w, *main_w, *sts_w, *count_w, *user_w;

void *thread(void *);

int sockfd;
int len;
int result;
int tmp;
int already_sigint;
bool away_mode;
string nickname;
string tmpaddr;
string hostname;
string msgs;
string msgr;
sockaddr_in address;
pthread_t my_thread;
char msg_s[MAX_DATA_SIZE];
char msg_r[MAX_DATA_SIZE];
color cl;
message ms;


void update_all()
{
    wnoutrefresh(main_w);
    wnoutrefresh(input_w);
    wnoutrefresh(sts_w);
    wnoutrefresh(count_w);
    wclear(input_w);
    doupdate();
}

void update_main()
{
    wnoutrefresh(main_w);
    doupdate();
    wrefresh(input_w);
}

void gui_init()
{
    initscr();
    main_w = newwin(LINES - 2, COLS - 21, 1, 0);
    sts_w = newwin(1, COLS - 21, 0, 0);
    count_w = newwin(1, 20, 0, COLS - 20);
    input_w = newwin(1, COLS, LINES - 1, 0);
    user_w = newwin(LINES - 2, 20, 1, COLS - 20);

    scrollok(main_w, TRUE);
    scrollok(input_w, FALSE);
    keypad(input_w, TRUE);

    update_all();
}

void color()
{
    if(!has_colors())
    {
	endwin();
	cerr << "ntalk: ncurses: has_colors() failed" << endl;
	cerr << "This terminal doesn't support colors. Try to use different." << endl;
	exit(1);
    }

    if(start_color() != OK)
    {
	endwin();
	cerr << "ntalk: ncurses: start_color() failed" << endl;
	cerr << "Ncurses could not start its color engine. Try to use different terminal." << endl;
	exit(2);
    }

    init_pair(1, COLOR_YELLOW, COLOR_BLACK);
    init_pair(2, COLOR_WHITE, COLOR_BLACK);
    init_pair(3, COLOR_WHITE, COLOR_BLUE);
    init_pair(4, COLOR_GREEN, COLOR_BLACK);
    init_pair(5, COLOR_WHITE, COLOR_RED);
}

void read_config()
{
    ifstream cfile;
    string filename;
    passwd *pwd_entry;
    char buf[255];

    pwd_entry = getpwuid(getuid());
    filename = pwd_entry->pw_dir;
    filename.append("/.ntalkrc");

    cfile.open(filename.c_str());
    if(!cfile)
    {
	cerr << "ntalk: " << filename << ": Could not open configuration file for reading." << endl;
	cerr << "ntalk: Using defaults." << endl;
	return;
    }

    while(!cfile.eof())
    {
	cfile.getline(buf, 255);
	string str = buf;

	if(str[0] == '#') continue;
	else if(str[0] == 'N')
	{
	    str.erase(0, 2);
	    nickname = str;
	}
	else if(str[0] == 'H')
	{
	    str.erase(0, 2);
	    tmpaddr = str;
	}
	else if(str[0] == 'C')
	{
	    str.erase(0, 2);
	    if(str[0] == 'Y')
	    {
		str.erase(0, 2);
		cl.you = atoi(str.c_str());
	    }
	    else if(str[0] == 'O')
	    {
		str.erase(0, 2);
		cl.other = atoi(str.c_str());
	    }
	    else if(str[0] == 'N')
	    {
		str.erase(0, 2);
		cl.nick = atoi(str.c_str());
	    }
	    else if(str[0] == 'B')
	    {
		str.erase(0, 2);
		cl.bracket = atoi(str.c_str());
	    }
	    else if(str[0] == 'S')
	    {
		str.erase(0, 2);
		cl.status = atoi(str.c_str());
	    }
	}
	else if(str[0] == 'M')
	{
	    str.erase(0, 2);
	    if(str[0] == 'L')
	    {
		str.erase(0, 2);
		ms.leave = str;
	    }
	    else if(str[0] == 'A')
	    {
		str.erase(0, 2);
		ms.away = str;
	    }
	}
    }

    cfile.close();
}

void save_config()
{
    fstream cfile, newcfile;
    string filename, newfilename, str;
    passwd *pwd_entry;
    char buf[255];

    pwd_entry = getpwuid(getuid());
    filename = pwd_entry->pw_dir;
    filename.append("/.ntalkrc");
    newfilename = "/tmp/ntalkrc_tmp";

    cfile.open(filename.c_str(), ios::in);
    if(!cfile)
    {
	cerr << "ntalk: "<< filename << ": Could not open configuration file for reading." << endl;
	return;
    }

    newcfile.open(newfilename.c_str(), ios::out);
    if(!newcfile)
    {
	cerr << "ntalk: " << newfilename << ": Could not open file for writing." << endl;
	return;
    }

    while(!cfile.eof())
    {
	strcpy(buf, "");
	cfile.getline(buf, 255);

	switch(buf[0]) {
	    case '#':
		newcfile << buf << endl;
		break;
	    case 'N':
		newcfile << "N=" << nickname << endl;
		break;
	    case 'H':
		newcfile << "H=" << tmpaddr << endl;
		break;
	    case 'C':
		if(buf[2] == 'Y')
		    newcfile << "C:Y=" << cl.you << endl;
		else if(buf[2] == 'O')
		    newcfile << "C:O=" << cl.other << endl;
		else if(buf[2] == 'N')
		    newcfile << "C:N=" << cl.nick << endl;
		else if(buf[2] == 'B')
		    newcfile << "C:B=" << cl.bracket << endl;
		else if(buf[2] == 'S')
		    newcfile << "C:S=" << cl.status << endl;
		break;
	    case 'M':
		if(buf[2] == 'L')
		    newcfile << "M:L=" << ms.leave << endl;
		else if(buf[2] == 'A')
		    newcfile << "M:A=" << ms.away << endl;
		break;
	    default:
		newcfile << buf << endl;
	}
    }

    cfile.close();
    newcfile.close();

    cfile.open(filename.c_str(), ios::out);
    newcfile.open(newfilename.c_str(), ios::in);

    while(!newcfile.eof())
    {
	newcfile.getline(buf, 255);
	cfile << buf << endl;
    }

    newcfile.close();
    cfile.close();
    unlink(newfilename.c_str());
}

string createStr(string com, string data, string to = "all")
{
    string m;
    m = nickname;
    m.append("@");
    m.append(hostname + ":");
    m.append(com + ":");
    m.append(to + ":");
    m.append(data);
    m.append("\n");

    return m;
}

void away_off()
{
    msgs = createStr(AWAYOFF, "I'm back.");
    strcpy(msg_s, msgs.c_str());

    write(sockfd, &msg_s, sizeof(msg_s));
    away_mode = 0;
    nocbreak();
    echo();
    mvwprintw(sts_w, 0, 0, "ntalk - user: %s\n", nickname.c_str());
    wrefresh(sts_w);
}

void refresh_user_w(const string messg)
{
    string msg = messg;
    int a, x;

    wclear(user_w);
    wrefresh(user_w);

    a = count(msg.begin(), msg.end(), ';');
    for(x = 0; x < a; x++)
    {
	// msg.replace(msg.find(';'), 1, '\n');// ntalk-client-0.6.3 originally used this, when I tried to compile it gave me this error:
	// ntalk.cpp: In function `void refresh_user_w(std::string)':
	// ntalk.cpp:345: error: invalid conversion from `char' to `const char*'
	// ntalk.cpp:345: error:   initializing argument 3 of `std::basic_string<_CharT, _Traits, _Alloc>& std::basic_string<_CharT, _Traits, _Alloc>::replace(typename _Alloc::size_type, typename _Alloc::size_type, const _CharT*) [with _CharT = char, _Traits = std::char_traits<char>, _Alloc = std::allocator<char>]'
	// make: *** [ntalk] Error 1
	// 
	// Added tempcstr for ntalk-client-1.0.0 and then changed the replace line (hope its right)
	// This removes the compile error but I'm not sure if it is the right way to fix this.
	char tempcstr[2];
	tempcstr[0] = '\n';
	tempcstr[1] = '\0';
	msg.replace(msg.find(';'), 1, tempcstr);
    }

    mvwprintw(user_w, 0, 0, "%s", msg.c_str());
    wrefresh(user_w);
}

void parse_user_command(const char *com)
{
    string _com = com;
    string c = _com.substr(0, _com.find(' '));
    string arg = _com.erase(0, _com.find(' ') + 1);

    wattron(sts_w, COLOR_PAIR(cl.status));

    if(c.compare("/q") == 0)
    {
	msgs = createStr(LEAVE, ms.leave);
	strcpy(msg_s, msgs.c_str());
	write(sockfd, &msg_s, sizeof(msg_s));
	endwin();
	save_config();
	exit(0);
    }
    else if(c.compare("/help") == 0)
    {
	mvwprintw(sts_w, 0, 0, "See README for more information\n");
	wrefresh(sts_w);
	wattron(main_w, COLOR_PAIR(2));
	wprintw(main_w, "-----------------------------------------------\n");
	wprintw(main_w, "Edit your ~/.ntalkrc to change default messages\n");
	wprintw(main_w, "and color schemes.\n");
	wprintw(main_w, "-----------------------------------------------\n");
	wprintw(main_w, "/help         - command list\n");
	wprintw(main_w, "/server       - information about server\n");
	wprintw(main_w, "/clear        - clear main window\n");
	wprintw(main_w, "/sclear       - clear status bar\n");
	wprintw(main_w, "/away         - switch to 'away' mode\n");
	wprintw(main_w, "/online       - refresh users box\n");
	wprintw(main_w, "/whois [nick] - simple information about user\n");
	wprintw(main_w, "/about        - about ntalk\n");
	wprintw(main_w, "/q            - quit ntalk\n");
	wprintw(main_w, "-----------------------------------------------\n");
	wnoutrefresh(main_w);
	doupdate();
    }
    else if(c.compare("/online") == 0)
    {
	msgs = createStr(ONLINE, "");
	strcpy(msg_s, msgs.c_str());
	write(sockfd, &msg_s, sizeof(msg_s));
    }
    else if(c.compare("/away") == 0)
    {
	msgs = createStr(AWAY, ms.away);
	strcpy(msg_s, msgs.c_str());
	write(sockfd, &msg_s, sizeof(msg_s));

	mvwprintw(sts_w, 0, 0, "Press 'x' to exit away mode.\n");
	wrefresh(sts_w);
	away_mode = 1;
	cbreak();
	noecho();
	while(away_mode == 1)
	{
	    tmp = wgetch(input_w);
	    if(tmp == 'x') away_off();
	}
    }
    else if(c.compare("/clear") == 0)
    {
	wclear(main_w);
	update_main();
    }
    else if(c.compare("/sclear") == 0)
    {
	mvwprintw(sts_w, 0, 0, "ntalk %s\n", VERSION);
	wrefresh(sts_w);
    }
    else if(c.compare("/about") == 0)
    {
	mvwprintw(sts_w, 0, 0, "Ncurses Talk, version %s by Krzysztof Czuba <kab@interia.pl>\n", VERSION);
	wrefresh(sts_w);
    }
    else if(c.compare("/server") == 0)
    {
	msgs = createStr(SERVER, "");
	strcpy(msg_s, msgs.c_str());
	write(sockfd, &msg_s, sizeof(msg_s));
    }
    else if(c.compare("/whois") == 0)
    {
	if(c.compare(arg) == 0)
	{
	    mvwprintw(sts_w, 0, 0, "Command usage: /whois [nick]\n");
	    wrefresh(sts_w);
	    return;
	}

	msgs = createStr(WHOIS, arg);
	strcpy(msg_s, msgs.c_str());
	write(sockfd, &msg_s, sizeof(msg_s));
    }
    else
    {
	wattron(sts_w, COLOR_PAIR(cl.status));
	mvwprintw(sts_w, 0, 0, "No such command: %s\n", c.c_str());
	wrefresh(sts_w);
    }
}

void parse_server_command(const string messg)
{
    string msg = messg;
    const string c_msg = messg;
    string nick, addr, com, to, data;

    nick = msg.substr(0, msg.find('@'));
    msg.erase(0, msg.find('@') + 1);
    addr = msg.substr(0, msg.find(':'));
    msg.erase(0, msg.find(':') + 1);
    com = msg.substr(0, msg.find(':'));
    msg.erase(0, msg.find(':') + 1);
    to  = msg.substr(0, msg.find(':'));
    msg.erase(0, msg.find(':') + 1);
    data = msg.substr(0, msg.find('\n'));

    wattron(main_w, COLOR_PAIR(2));

    if(com.compare(MESSAGE) == 0)
    {
	wattron(main_w, COLOR_PAIR(cl.bracket));
	wprintw(main_w, "<");
	wattron(main_w, COLOR_PAIR(cl.nick));
	wprintw(main_w, "%s", nick.c_str());
	wattron(main_w, COLOR_PAIR(cl.bracket));
	wprintw(main_w, ">");
	if(nick.compare(nickname) == 0) wattron(main_w, COLOR_PAIR(cl.you));
	else wattron(main_w, COLOR_PAIR(cl.other));
	wprintw(main_w, " %s\n", data.c_str());
	update_main();
    }
    else if(com.compare(JOIN) == 0)
    {
	//wprintw(main_w, "---> %s@%s has joined the room\n", nick.c_str(), addr.c_str());
	wprintw(main_w, "---> %s joined the room\n", nick.c_str());
	update_main();

	msgs = createStr(ONLINE, "");
	strcpy(msg_s, msgs.c_str());
	write(sockfd, &msg_s, sizeof(msg_s));
    }
    else if(com.compare(LEAVE) == 0)
    {
	wprintw(main_w, "---> %s left the room (%s)\n", nick.c_str(), data.c_str());
	update_main();
    }
    else if(com.compare(SERVER) == 0)
    {
	wprintw(main_w, "---> %s\n", data.c_str());
	update_main();
    }
    else if(com.compare(AWAY) == 0)
    {
	wprintw(main_w, "---> %s is away from computer: %s\n", nick.c_str(), data.c_str());
	update_main();
    }
    else if(com.compare(AWAYOFF) == 0)
    {
	wprintw(main_w, "---> %s is back.\n", nick.c_str());
	update_main();
    }
    else if(com.compare(ONLINE) == 0)
    {
	refresh_user_w(data);
    }
    else if(com.compare(DISCONNECT) == 0)
    {
	endwin();
	cout << "\nDisconnected: " << data << endl;
	exit(0);
    }
    else if(com.compare(WHOIS) == 0)
    {
	wprintw(main_w, "---> %s is: %s@%s (since %s)\n", nick.c_str(), nick.c_str(), addr.c_str(), data.c_str());
	update_main();
    }
    else if(com.compare(SRVMESSAGE) == 0)
    {
	wprintw(main_w, "---> %s\n", data.c_str());
	update_main();
    }
    else if(com.compare(SHUTDOWN) == 0)
    {
	save_config();
	endwin();
	cout << "ntalk: Server shutdown" << endl;
	exit(0);
    }
}

string get_host(string hostname)
{
    char **addr;
    hostent *host_info;

    host_info = gethostbyname(hostname.c_str());
    if(!host_info)
    {
	cerr << "ntalk: gethostbyname() failed" << endl;
	exit(1);
    }

    addr = host_info->h_addr_list;
    string str = inet_ntoa(*(in_addr *)*addr);

    return str;
}

static void signal_interrupt(int sig)
{
    msgs = createStr(LEAVE, "Interrupted...");
    strcpy(msg_s, msgs.c_str());

    if(!already_sigint)
    {
	already_sigint = 1;
	write(sockfd, &msg_s, sizeof(msg_s));
	endwin();
	exit(0);
    }
}

int main(int argc, char *argv[])
{
    int opt;
    passwd *pwd_entry;
    char hostn[256];

    // Some defaults...

    tmpaddr = "127.0.0.1";
    pwd_entry = getpwuid(getuid());
    nickname = pwd_entry->pw_name;

    cl.you 	= 1;
    cl.other 	= 2;
    cl.nick 	= 2;
    cl.bracket 	= 4;
    cl.status   = 3;

    ms.leave = "Leaving...";
    ms.away = "Busy...";

    read_config();

    while((opt = getopt(argc, argv, "n:h:v?")) != -1)
    {
	if(opt == 'h')
	{
		if(strlen(optarg) < 1)
		{
		cout << "hostname requires an argument!\n";
		exit(0);
	    }
		tmpaddr = optarg;
	}
	else if(opt == 'n')
	{
	    if((strlen(optarg) > 21) || (strlen(optarg) < 2))
	    {
		cout << "Error: Your nick must contain 2 - 20 charackters\n";
		exit(0);
	    }
	    nickname = optarg;
	}
	else if(opt == 'v')
	{
	    cout << "version " << VERSION << " (ncurses 5.2)\n(C) Krzysztof Czuba <kab@interia.pl>\n";
	    exit(0);
	}
	else if(opt == '?')
	{
	    cout << "\nUsage: client [option]\n\nOptions:\n"
                   "-h\t\tspecify server address\n"
		   "-n\t\tspecify nickname\n"
                   "-v\t\tshow version information\n"
		   "-?\t\tdisplay this information\n\n";
	    exit(0);
	}
	else exit(0);
    }

    if(gethostname(hostn, 255) == -1)
    {
	cerr << "ntalk: gethostname() failed" << endl;
	cerr << "Ntalk could not obtain your computer's name." << endl;
	exit(0);
    }

    hostname = hostn;

    tmpaddr = get_host(tmpaddr);

    sockfd = socket(AF_INET, SOCK_STREAM, 0);
    address.sin_family = AF_INET;
    address.sin_addr.s_addr = inet_addr(tmpaddr.c_str());
    address.sin_port = htons(9734);
    len = sizeof(address);
    result = connect(sockfd, (sockaddr *)&address, len);

    if(result == -1)
    {
	endwin();
	cerr << "ntalk: connect() failed" << endl;
	cerr << "ntalk: Make sure that the address is valid and the server is running" << endl;
	exit(1);
    }

    gui_init();
    color();

    wattron(sts_w, COLOR_PAIR(cl.status));
    wattron(count_w, COLOR_PAIR(3));
    mvwprintw(sts_w, 0, 0, "ntalk %s - user: %s | type /help for command list\n", VERSION, nickname.c_str());
    mvwprintw(count_w, 0, 0, "      Online:      \n");
    wrefresh(sts_w);
    wrefresh(count_w);

    msgs = createStr(JOIN, "");
    strcpy(msg_s, msgs.c_str());
    write(sockfd, &msg_s, sizeof(msg_s));

    already_sigint = 0;
    signal(SIGINT, signal_interrupt);

    if(pthread_create(&my_thread, NULL, thread, (void *)"watek") != 0)
    {
	cerr << "ntalk: pthread_create() failed" << endl;
	cerr << "Could not create thread." << endl;
	exit(1);
    }

    while(1)
    {
	char input[MAX_DATA_SIZE];
	wgetnstr(input_w, input, MAX_DATA_SIZE);
	input[MAX_DATA_SIZE - 1] = '\0';
	if(input[0] == '/')
	{
	    parse_user_command(input);
	    wprintw(input_w, "\n");
	    continue;
	}

	msgs = createStr(MESSAGE, string(input));
	strcpy(msg_s, msgs.c_str());

	write(sockfd, &msg_s, sizeof(msg_s));
	wprintw(input_w, "\n");
    }

    close(sockfd);
    exit(0);
}

void *thread(void *watek)
{
    while(1)
    {
	read(sockfd, &msg_r, sizeof(msg_r));
	string msg = msg_r;
	parse_server_command(msg);
    }
}
